compiler-rt/lib/asan/asan_thread.cpp

```diff
@@ -93,6 +97,11 @@ AsanThreadContext *GetThreadContextByTidLocked(u32 tid) {
 AsanThread *AsanThread::Create(const void *start_data, uptr data_size,
                                u32 parent_tid, StackTrace *stack,
                                bool detached) {
+#if SANITIZER_WINDOWS
```

Comment:
how newthread can happen? Isn't there is interceptor for CreateThread which should be executed first on main thread?

Reply:
Looks like dead lock prone solution
Whoever is creator of this thread may wait to join this what, and prevent initialization happening

Reply:
From .NET, if a worker thread loads an ASAN instrumented binary, we can see all of the checks in the PR description fail based on timing.

For the check `((!asan_init_is_running && "ASan init calls itself!")) != (0)`, this can happen when one thread is initializing ASan and has already intercepted `malloc` or `Rtl*` APIs. While it's still initializing, another thread calls the intercepted API, which results in the `__asan::Allocator` making another call to `AsanInitInternal`, and the check fires.

For the main thread creation, this happens less frequently since the call to `CreateMainThread` is nearly at the end of `AsanInitInternal` shortly after `CreateThread` has been intercepted. After it's intercepted, the thread registry can attempt to create another thread before the call to create the main thread. I'm not sure how ASan can be dynamically loaded on other platforms besides Windows, so this might not be an issue outside of Windows.

Reply:
@vitalybuka, do the changes look sufficient to you after the latest push?

Reply:
You replied as "will do" to some items, as #if
But I see update in this PR.

Also there are https://github.com/zahiraam/llvm-project/commit/8e9ce62cf037767963ecd1a48aeb491bc0ef5863
you want to create PR for them?

BTW. There is https://github.com/getcord/spr which is proposed to be used for stacked code review, similar to we have in Phabricator, you may try that.

Reply:
I guess  just create PRs for 8ef594b and 8e9ce62
I will merge them asap, and you can rebase this one

Reply:
Apologies... I think I'm confused on how PRs work here. 😅 

Are you saying rebase this branch with those, and then merge? In the future, is there a document for me to reference as to not trouble you with this?

Reply:
I we don't have documentation yet, after migrating to pull request.

So easiest way is to create sevaral branches in your fork and create PR from each one:
1. format
2. format + asan_init_is_running
3. format + asan_init_is_running + the rest  (this one)

2 and 3 will have include some code from PRs above, but they will go away after merging earlier one and rebase

https://github.com/getcord/spr does this stuff for you, but it will required undocumented setup

Reply:
@vitalybuka is this what you had in mind? I tried using spr

https://github.com/llvm/llvm-project/pull/73819
https://github.com/llvm/llvm-project/pull/73822
https://github.com/llvm/llvm-project/pull/73824

Reply:
It's close, but I assume you have branch per PR locally, because PRs include diffs from other PRs

With SPR you can have all patches in local branch, and upload them with spr diff --all, then it will create stacked review similar we had with Phabricator.

Reply:
And SPR is suppose to upload into upstream repo, not your fork
`llvm-project/.git/config` should look like:

```
	githubRemoteName = origin
	githubRepository = llvm/llvm-project
	githubMasterBranch = main
	branchPrefix = users/vitalybuka/spr/
	requireTestPlan = false
```

Reply:
I assumed I had to use my fork to create the PR since I don't have push permissions to llvm. I tried changing the configuration and received invalid PR.

I'll see if @barcharcraz can help me with this live today.

Reply:
Sorry for that, I expected it would be easier.
We can go one by one with regular PRs.

Reply:
I was able to get LLVM push permissions, and I created the PRs using `spr diff --all` on the main repo. Here is the last stacked PR. https://github.com/llvm/llvm-project/pull/74086